lrpar/src/lib/ctbuilder.rs

```diff
@@ -1125,6 +1109,76 @@ where
     }
 }
 
+pub struct CTAnalysisBuilder<'a, LexemeT, StorageT = u32>
+where
+    StorageT: Eq + Hash,
+{
+    pb: CTParserBuilder<'a, LexemeT, StorageT>,
+}
+
+impl<'a, LexemeT, StorageT> CTAnalysisBuilder<'a, LexemeT, StorageT>
+where
+    LexemeT: Lexeme<StorageT>,
+    StorageT: 'static + Debug + Hash + PrimInt + Serialize + Unsigned,
+    usize: AsPrimitive<StorageT>,
+{
+    pub fn read_grammar(
+        self,
+        inc: &mut String,
+    ) -> Result<CTGrammarASTAnalyzerBuilder<'a, LexemeT, StorageT>, Box<dyn Error>> {
+        inc.clear();
+        let grmp = self
+            .pb
+            .grammar_path
+            .as_ref()
+            .expect("grammar_path must be specified before processing.");
+        let outp = self
+            .pb
+            .output_path
+            .as_ref()
+            .expect("output_path must be specified before processing.");
+        let yk = match self.pb.yacckind {
+            None => panic!("yacckind must be specified before processing."),
+            Some(YaccKind::Original(x)) => YaccKind::Original(x),
+            Some(YaccKind::Grmtools) => YaccKind::Grmtools,
+            Some(YaccKind::Eco) => panic!("Eco compile-time grammar generation not supported."),
+        };
+
+        {
+            let mut lk = GENERATED_PATHS.lock().unwrap();
+            if lk.contains(outp.as_path()) {
+                return Err(format!("Generating two parsers to the same path ('{}') is not allowed: use CTParserBuilder::output_path (and, optionally, CTParserBuilder::mod_name) to differentiate them.", &outp.to_str().unwrap()).into());
+            }
+            lk.insert(outp.clone());
+        }
+
+        let mut file = File::open(&grmp)?;
+        file.read_to_string(inc)?;
+        Ok(CTGrammarASTAnalyzerBuilder {
+            fmeta: FileMeta {
+                grmp: grmp.to_owned(),
+                outp: outp.to_owned(),
+                yk,
+            },
+            pb: self.pb,
+        })
+    }
+}
+
+struct FileMeta {
```

Comment:
What's the utility of this struct?

Reply:
Well, before I introduced this you could still get the data in *fmeta*, but it was stuck in one of the builders, so you would end up with `builder.tb.ab.pb` and `builder.tb.ab.grmp`, it seemed instead it would be better to just keep the `pb` and the various data that was still needed rather than keeping all the builder self values.